services/oidc/authenticator.js

```diff
@@ -0,0 +1,104 @@
+const OidcError = require('./error');
+
+class OidcAuthenticator {
+  /**
+   * @param {import('../../context/init').Context} context
+   */
+  constructor(context) {
+    /** @private @readonly */
+    this.openIdClient = context.openIdClient;
+    /** @private @readonly */
+    this.env = context.env;
+    /** @private @readonly */
+    this.process = context.process;
+
+    ['openIdClient', 'env', 'process'].forEach((name) => {
+      if (!this[name]) throw new Error(`Missing dependency ${name}`);
+    });
+  }
+
+  /**
+   * @private
+   * @returns {Promise<import('openid-client').Client>}
+   */
+  async register() {
+    try {
+      const issuer = await this.openIdClient.Issuer.discover(`${this.env.FOREST_URL}/oidc/.well-known/openid-configuration`);
+
+      return await issuer.Client.register({
+        name: 'lumber',
+        application_type: 'native',
+        redirect_uris: ['com.forestadmin.lumber://authenticate'],
+        token_endpoint_auth_method: 'none',
+        grant_types: ['urn:ietf:params:oauth:grant-type:device_code'],
+        response_types: ['none'],
+      });
+    } catch (e) {
+      throw new OidcError(
+        'Unable to register against the Forest Admin server',
+        e,
+      );
+    }
+  }
+
+  /**
+   * @private
+   * @param {import('openid-client').Client} client
+   * @returns {Promise<import('openid-client').DeviceFlowHandle>}
+   */
+  static async launchDeviceAuthorization(client) {
+    try {
+      return await client.deviceAuthorization({
+        scopes: ['openid', 'email', 'profile'],
+      });
+    } catch (e) {
+      throw new OidcError(
+        'Error while starting the authentication flow',
+        e,
+      );
+    }
+  }
+
+  /**
+   * @private
+   * @param {import('openid-client').DeviceFlowHandle} flow
+   * @returns {Promise<import('openid-client').TokenSet>}
+   */
+  async waitForAuthentication(flow) {
+    const expiresIn = flow.expires_in;
+    try {
+      this.process.stdout.write(`Navigate to the url: ${flow.verification_uri}\n`);
+      this.process.stdout.write(`and enter your verification code: ${flow.user_code}\n`);
```

Comment:
Why not use the logger here ?

Reply:
Because that does not correspond to logs, they are some messages for the user. Logs are appended to the console with the current configuration, but that's not necessarily always the case (depending on the config).

Here, I want to always display some content to the user.

I hesitated, but this way seems more clean. That's also why I did not use `console.log`, because it corresponds more to a stdout output even if the result is the same.